### PR TITLE
Time indexes for detector table

### LINKED ISSUE #19380: Add Time Indexes to Detector Table
To make testing for scanning workspaces a little easier time index information should be added to the detector table in Mantid. The behaviour should be similar to that for detector IDs, with multiple time indexes being shown where appropriate.

### PR DESCRIPTION
**Pull request by ianbush**
This adds a column showing the time indexes to the detector table. The logic is the same as for detector IDs - if there are more than 10 entries then the indexes are just displayed as `1, 2...(100 more)...102, 103`.

**To test:**

Please use the changes in #19728 to test this branch. `CreateSampleWorkspace` now has the option to create a workspace with mulitple time indexes, and `GroupDetectors` can be used to test workspaces with more than one time index.

1. A workspace with no time indexing should not show the column at all
1. A workspace with less a single time index should show the time index
1. A workspace with less than 10 time indexes should show in a comma separated list
1. A workspace with more than 10 time indexes should show in a list such as `1, 2...(100 more)...102, 103`

Fixes #19380.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Types/CMakeLists.txt [hunk 1]
```diff
@@ -14,5 +14,11 @@ if (COVERALLS)
   endforeach(loop_var)
 endif()
 
+# Add the target for this directory
+add_library ( Types ${INC_FILES} )
+# Set the name of the generated library
+set_target_properties ( Types PROPERTIES OUTPUT_NAME MantidTypes
+        COMPILE_DEFINITIONS IN_MANTID_TYPES LINKER_LANGUAGE CXX )
```
Comment (SimonHeybrock) [inline Framework/Types/CMakeLists.txt]: `Types` is a header-only module so I am not sure why this should be required...?

File: MantidPlot/src/Mantid/MantidUI.cpp [hunk 2]
```diff
@@ -1377,6 +1361,15 @@ Table *MantidUI::createDetectorTable(
...
       colValues << QVariant(specNo) << QVariant(detIds);
+      if (isScanning) {
+        std::set<int> timeIndexSet;
+        for (auto def : spectrumInfo.spectrumDefinition(wsIndex)) {
+          timeIndexSet.insert(int(def.second));
+        }
+
+        QString timeIndexes = createTruncatedList(timeIndexSet);
+        colValues << QVariant(timeIndexes);
+      }
```
Comment (SimonHeybrock) [inline MantidPlot/src/Mantid/MantidUI.cpp]: So if there is more than one detector for the spectrum you would get duplicate time indices and cannot associate them to detectors? Probably this is not a common case, but just to make sure this is not overlooked. Should it simply be ignored? Or generate a warning? Throw?

Comment (ianbush) [inline MantidPlot/src/Mantid/MantidUI.cpp]: Just to be clear, are you saying that you could have a spectrum with 2 detectors, one contributing time indexes [1, 2] and another contributing time indexes [2, 3]. You would then see the time indexes as [1, 2, 3] without knowing which detectors had which time indexes?

I would think we can ignore that, as it should be easy to see when you have multiple time indexes and multiple detectors. The detector table does not always give the complete information in any case, truncating long lists.

The other alternative would be to do something like "1 (1, 2), "2 (2, 3)" in the detector part of the table. I think that would get more difficult to read clearly though.

Comment (SimonHeybrock) [inline MantidPlot/src/Mantid/MantidUI.cpp]: Yes that is what I meant. I suppose you would see the time indices as [1,2,2,3].

Comment (ianbush) [inline MantidPlot/src/Mantid/MantidUI.cpp]: But it is coming from a set, so you would see [1,2,3] ?

File: Framework/Types/CMakeLists.txt (hunk 1 above)
Comment (ianbush) [inline Framework/Types/CMakeLists.txt]: Is this what you would expect?

Comment (SimonHeybrock) [inline Framework/Types/CMakeLists.txt]: I am not sure. Why are you even creating a library? Maybe making it an interface library would make sense? ```add_library( Types INTERFACE ...)```

Comment (ianbush) [inline Framework/Types/CMakeLists.txt]: `add_library ( Types INTERFACE )` seems to be sufficent for the build to work. Without this I get the error: `cannot find -lTypes`.

**Comment by SimonHeybrock** [comment]
Looks good now, :shipit:

**Pull request merged**